Adds confirms to scaffold

kevindoole opened the pull request:
For issue #1797 

Uses `WP_CLI::confirm()` so there's a `--yes` flag.

Output looks like this:
```
Warning: Scaffold will overwrite: 
/Users/kevindoole/code/htdocs/agcanada/wp-content/plugins/blah/blah.php
/Users/kevindoole/code/htdocs/agcanada/wp-content/plugins/blah/readme.txt
/Users/kevindoole/code/htdocs/agcanada/wp-content/plugins/blah/package.json
/Users/kevindoole/code/htdocs/agcanada/wp-content/plugins/blah/Gruntfile.js
Overwrite files and continue? [y/n] y
Success: Created /Users/kevindoole/code/htdocs/agcanada/wp-content/plugins/blah
Warning: Scaffold will overwrite: 
/Users/kevindoole/code/htdocs/agcanada/wp-content/plugins/blah/tests/bootstrap.php
Overwrite files and continue? [y/n] y
Success: Created test files.
```

In the first commit, it warned about overwriting files one at a time, and saying "no" would cause the process to exit. So, taking the above example, if you said yes to 'blah.php' then no to 'readme.txt' the process would exit without even trying the remaining files.

I think grouping the files is a little better so you get to see everything that will get overwritten at once. The one exception, shown above, is for `wp scaffold plugin` where the plugin tests generate a separate rewrite warning. Seems ok.

Thoughts?

kevindoole commented:
(Aack, somehow didn't notice the [original pull request](https://github.com/wp-cli/wp-cli/pull/1088) for this one yesterday. fwiw, the discussion there leans toward a file-by-file confirmation.)

danielbachhuber commented:
> I think grouping the files is a little better so you get to see everything that will get overwritten at once.

I think my preference would be one by one, so you could opt-in or opt out for each one.

> So, taking the above example, if you said yes to 'blah.php' then no to 'readme.txt' the process would exit without even trying the remaining files.

You could roll your own with `cli\Streams\Prompt`: https://github.com/wp-cli/php-cli-tools/blob/bfe6d35911ecebef2edea7db3c9de9a2b74aa6a8/lib/cli/Streams.php#L145

Can you implement with `--force` instead of `--yes` ? `--force` is more semantically correct, because you're overwriting original files.

kevindoole commented:
Agreed. So, something like:

```
$ wp scaffold plugin blah

File already exists:
/Users/kevindoole/code/htdocs/agcanada/wp-content/plugins/blah/blah.php
Skip this file, or replace it with scaffolding? [s/r] r

File already exists
/Users/kevindoole/code/htdocs/agcanada/wp-content/plugins/blah/readme.txt
Skip this file, or replace it with scaffolding? [s/r] s

File already exists
/Users/kevindoole/code/htdocs/agcanada/wp-content/plugins/blah/package.json
Skip this file, or replace it with scaffolding? [s/r] s

File already exists
/Users/kevindoole/code/htdocs/agcanada/wp-content/plugins/blah/Gruntfile.js
Skip this file, or replace it with scaffolding? [s/r] r

Success: Created /Users/kevindoole/code/htdocs/agcanada/wp-content/plugins/blah

File already exists
/Users/kevindoole/code/htdocs/agcanada/wp-content/plugins/blah/tests/bootstrap.php
Skip this file, or replace it with scaffolding? [s/r] r

Success: Created test files.

$ wp scaffold plugin blah --force

Success: Created /Users/kevindoole/code/htdocs/agcanada/wp-content/plugins/blah
Success: Created test files.
```

danielbachhuber commented:
> So, something like:

Yep. A neat additional enhancement (not required in this PR) would be to run the files through `diff` so you can see the visual difference before merging.

kevindoole commented:
Totally! I'll keep it in mind this afternoon when I make these changes -- could be simple to include.

kevindoole commented:
I think i've covered everything here at this point, although i wrote no tests. Seems like this should be tested, no?

(`WP_CLI@prompt` is pretty rickety -- just mocked it up quick in order to prepare this output.)

In any case, here are some examples. Long story short, it prompts when a file will be overwritten. At the end of the process, it gives a success message when files are written, or a message that all files were skipped. There's the option of, along with the success message, additionally showing which files were created.

```
$ wp scaffold taxonomy test --theme=twentythirteen
Success: Created /Users/kevindoole/code/htdocs/wptests/wp-content/themes/twentythirteen/taxonomies/testing.php
$ wp scaffold taxonomy test --theme=twentythirteen
Warning: File already exists:
/Users/kevindoole/code/htdocs/wptests/wp-content/themes/twentythirteen/taxonomies/testing.php
Skip this file, or replace it with scaffolding? [s/r] s


Skipped creating /Users/kevindoole/code/htdocs/wptests/wp-content/themes/twentythirteen/taxonomies/testing.php
$ wp scaffold taxonomy test --theme=twentythirteen
Warning: File already exists:
/Users/kevindoole/code/htdocs/wptests/wp-content/themes/twentythirteen/taxonomies/testing.php
Skip this file, or replace it with scaffolding? [s/r] r


Success: Created /Users/kevindoole/code/htdocs/wptests/wp-content/themes/twentythirteen/taxonomies/testing.php
```

```
$ wp scaffold plugin testing
Success: Created plugin files.
Success: Created test files.
$ wp scaffold plugin testing
Warning: File already exists:
/Users/kevindoole/code/htdocs/wptests/wp-content/plugins/testing/testing.php
Skip this file, or replace it with scaffolding? [s/r] s


Warning: File already exists:
/Users/kevindoole/code/htdocs/wptests/wp-content/plugins/testing/readme.txt
Skip this file, or replace it with scaffolding? [s/r] s


Warning: File already exists:
/Users/kevindoole/code/htdocs/wptests/wp-content/plugins/testing/package.json
Skip this file, or replace it with scaffolding? [s/r] s


Warning: File already exists:
/Users/kevindoole/code/htdocs/wptests/wp-content/plugins/testing/Gruntfile.js
Skip this file, or replace it with scaffolding? [s/r] s


All plugin files were skipped.
Warning: File already exists:
/Users/kevindoole/code/htdocs/wptests/wp-content/plugins/testing/tests/bootstrap.php
Skip this file, or replace it with scaffolding? [s/r] s


Warning: File already exists:
/Users/kevindoole/code/htdocs/wptests/wp-content/plugins/testing/bin/install-wp-tests.sh
Skip this file, or replace it with scaffolding? [s/r] s


Warning: File already exists:
/Users/kevindoole/code/htdocs/wptests/wp-content/plugins/testing/.travis.yml
Skip this file, or replace it with scaffolding? [s/r] s


Warning: File already exists:
/Users/kevindoole/code/htdocs/wptests/wp-content/plugins/testing/phpunit.xml
Skip this file, or replace it with scaffolding? [s/r] s


Warning: File already exists:
/Users/kevindoole/code/htdocs/wptests/wp-content/plugins/testing/tests/test-sample.php
Skip this file, or replace it with scaffolding? [s/r] s


All test files were skipped.
$ wp scaffold plugin testing
Warning: File already exists:
/Users/kevindoole/code/htdocs/wptests/wp-content/plugins/testing/testing.php
Skip this file, or replace it with scaffolding? [s/r] s


Warning: File already exists:
/Users/kevindoole/code/htdocs/wptests/wp-content/plugins/testing/readme.txt
Skip this file, or replace it with scaffolding? [s/r] r


Warning: File already exists:
/Users/kevindoole/code/htdocs/wptests/wp-content/plugins/testing/package.json
Skip this file, or replace it with scaffolding? [s/r] r


Warning: File already exists:
/Users/kevindoole/code/htdocs/wptests/wp-content/plugins/testing/Gruntfile.js
Skip this file, or replace it with scaffolding? [s/r] s


Success: Created plugin files.
Warning: File already exists:
/Users/kevindoole/code/htdocs/wptests/wp-content/plugins/testing/tests/bootstrap.php
Skip this file, or replace it with scaffolding? [s/r] r


Warning: File already exists:
/Users/kevindoole/code/htdocs/wptests/wp-content/plugins/testing/bin/install-wp-tests.sh
Skip this file, or replace it with scaffolding? [s/r] as


Warning: File already exists:
/Users/kevindoole/code/htdocs/wptests/wp-content/plugins/testing/.travis.yml
Skip this file, or replace it with scaffolding? [s/r] s


Warning: File already exists:
/Users/kevindoole/code/htdocs/wptests/wp-content/plugins/testing/phpunit.xml
Skip this file, or replace it with scaffolding? [s/r] r


Warning: File already exists:
/Users/kevindoole/code/htdocs/wptests/wp-content/plugins/testing/tests/test-sample.php
Skip this file, or replace it with scaffolding? [s/r] s


Success: Created test files.
```

danielbachhuber commented:
> Seems like this should be tested, no?

Yes please. You can test by including `--force` param in your test, and then watch `STDOUT` for "Warning.... Replaced"

kevindoole commented:
Tests added, --force works correctly. The output has changed slightly:
```
$ wp scaffold plugin test
Warning: /Users/kevindoole/code/htdocs/wptests/wp-content/plugins/test/test.php already exists.
Skip this file, or replace it with scaffolding?[s/r]: s
Skipped
Warning: /Users/kevindoole/code/htdocs/wptests/wp-content/plugins/test/readme.txt already exists.
Skip this file, or replace it with scaffolding?[s/r]: s
Skipped
Warning: /Users/kevindoole/code/htdocs/wptests/wp-content/plugins/test/package.json already exists.
Skip this file, or replace it with scaffolding?[s/r]: r
Replaced
Warning: /Users/kevindoole/code/htdocs/wptests/wp-content/plugins/test/Gruntfile.js already exists.
Skip this file, or replace it with scaffolding?[s/r]: s
Skipped
Success: Created plugin files.
Warning: /Users/kevindoole/code/htdocs/wptests/wp-content/plugins/test/tests/bootstrap.php already exists.
Skip this file, or replace it with scaffolding?[s/r]: s
Skipped
Warning: /Users/kevindoole/code/htdocs/wptests/wp-content/plugins/test/bin/install-wp-tests.sh already exists.
Skip this file, or replace it with scaffolding?[s/r]: s
Skipped
Warning: /Users/kevindoole/code/htdocs/wptests/wp-content/plugins/test/.travis.yml already exists.
Skip this file, or replace it with scaffolding?[s/r]: s
Skipped
Warning: /Users/kevindoole/code/htdocs/wptests/wp-content/plugins/test/phpunit.xml already exists.
Skip this file, or replace it with scaffolding?[s/r]: r
Replaced
Warning: /Users/kevindoole/code/htdocs/wptests/wp-content/plugins/test/tests/test-sample.php already exists.
Skip this file, or replace it with scaffolding?[s/r]: r
Replaced
Success: Created test files.
```

```
$ wp scaffold plugin-tests test --force
Warning: /Users/kevindoole/code/htdocs/wptests/wp-content/plugins/test/tests/bootstrap.php already exists.
Replaced
Warning: /Users/kevindoole/code/htdocs/wptests/wp-content/plugins/test/bin/install-wp-tests.sh already exists.
Replaced
Warning: /Users/kevindoole/code/htdocs/wptests/wp-content/plugins/test/.travis.yml already exists.
Replaced
Warning: /Users/kevindoole/code/htdocs/wptests/wp-content/plugins/test/phpunit.xml already exists.
Replaced
Warning: /Users/kevindoole/code/htdocs/wptests/wp-content/plugins/test/tests/test-sample.php already exists.
Replaced
Success: Created test files.
```

danielbachhuber commented:
Looks like there's formatting issues?

kevindoole commented:
Ah, yeah; i was screwing around with it a bit -- fixed!

danielbachhuber commented on php/class-wp-cli.php:
```diff
@@ -287,6 +287,25 @@ public static function confirm( $question, $assoc_args = array() ) {
+		return $answers[ 1 ] === $answer;
+	}
+
+	/**
```
Er, I don't think we need this abstraction. Can you just include in your method?

danielbachhuber commented on features/scaffold.feature:
```diff
@@ -336,3 +336,17 @@ Feature: WordPress code scaffolding
+    And STDOUT should contain:
+    """
+    Replaced
+    """
```
Needs newline

kevindoole replied:
added

danielbachhuber commented on features/scaffold.feature:
```diff
@@ -336,3 +336,17 @@ Feature: WordPress code scaffolding
     Error: Could not decompress your theme files
     """
+
+  @wip
```
Please remove `@wip` flag

danielbachhuber commented on php/commands/scaffold.php:
```diff
@@ -408,19 +444,29 @@ public function package_tests( $args, $assoc_args ) {
+
+			$force = \WP_CLI\Utils\get_flag_value( $assoc_args, 'force' );
+			$should_write_file = $this->prompt_if_files_will_be_overwritten( $file_path, $force );
+			if ( ! $should_write_file ) continue;
```
Newline, and wrap the conditional with brackets, per coding standards.

danielbachhuber commented on php/commands/scaffold.php:
```diff
@@ -558,24 +619,73 @@ function plugin_tests( $args, $assoc_args ) {
+		return $wrote_files;
+	}
+
+	private function prompt_if_files_will_be_overwritten( $filename, $force ) {
```
I'm not sure how I feel about this method.

danielbachhuber commented:
Some small nits. I'm also not 100% on the abstraction right now - need to mull it over.

kevindoole commented:
Ok, fixed everything, but left `prompt_if_files_will_be_overwritten()` as is for now.

danielbachhuber commented:
> Ok, fixed everything, but left prompt_if_files_will_be_overwritten() as is for now.

Eh, it's a private method. We can always change it later.
